Describe how the input source list is maintained

toji opened the pull request:
Fixes #465.

Defines what should happen as input sources for a session are added,
removed, and changed. Also clarifies the definitions for some of the
related concepts.

Manishearth commented on index.bs:
```diff
@@ -475,11 +475,47 @@ When the <dfn method for="XRSession">requestReferenceSpace(|options|)</dfn> meth
+
+  1. Let |oldInputSource| be the {{XRInputSource}} in |session|'s [=list of active XR input sources=] previously associated with the [=XR input source=] that changed.
+  1. Let |newInputSource| be a new {{XRInputSource}}.
+  1. Remove |oldInputSource| from |session|'s [=list of active XR input sources=].
```
Should we also unset an `active` flag on the source? Make `getPose()` on inactive input source spaces returns null (or throw an error for better diagnostics?).

In other words, force people to reload input source lists whenever they change. Otherwise they'll continue using a nonexistant input space.

toji replied:
Good suggestion! I'll incorporate that.

toji replied:
Following up: This is ever-so-slightly more complex than I had originally anticipated, primarily because I think the "active" flag should actually reside on the space and not the input source. That implies that we probably want to make more consistent use of it elsewhere in the spec, so I wanted to give it some more thought before jumping into a change like that.

Manishearth replied:
Totally okay with it being a followup, just something to keep in mind.

toji commented:
Pushed some changes to this PR, thanks for your patience!

> Is there a reason this PR isn't reflecting the behavior of the inputsourceschange event...

Huh. Turns out that the spec was missing the XRInputSourceChangeEvent entirely! Fixed that now, as well as making sure the algorithms populate the `added`/`removed` arrays as needed.

One minor concern: The event name is  "inputsource**s**change" while the interface, as defined in the explainer, is "XRInputSourceChangeEvent" (note the singular "Source"). We should probably normalize that, and it could easily be done in this PR. I suggest making the interface use the plural form?

> Is the expectation that left/right motion controllers will each fire their own event even if they're detected on the same frame?

Nope, that was an oversight on my part. Fixed in the latest commit.

> Are there any developer expectations we should be setting about when input sources may change?

We certainly can, though what exactly those expectations should be is not readily apparent to me at the moment and is probably best discussed in a separate issue.

NellWaliczek commented:
Looking at the other open PRs made me realize that one really important thing is missing.  We haven't defined the expected behavior of the array when an item is added/removed.  Do gaps fill in?  What about for transient input sources?

Manishearth commented:
> Looking at the other open PRs made me realize that one really important thing is missing. We haven't defined the expected behavior of the array when an item is added/removed. Do gaps fill in? What about for transient input sources?

@NellWaliczek There's some discussion on that in #624. It's not touching array ordering concerns, though.

cwilso commented:
Also fixes #632

toji commented on index.bs:
```diff
@@ -486,33 +486,44 @@ The user agent MUST monitor any [=XR input source=]s associated with the [=/XR D
+  1. For each new [=XR input source=]:
+    1. Let |inputSource| be a new {{XRInputSource}}.
+    1. Add |inputSource| to |added|.
+    1. Add |inputSource| to |session|'s [=list of active XR input sources=] associated with the [=XR input source=].
```
Ooh! Good catch! Thanks!

toji commented on index.bs:
```diff
@@ -486,33 +486,44 @@ The user agent MUST monitor any [=XR input source=]s associated with the [=/XR D
 <div class="algorithm" data-algorithm="on-input-source-change">
 
-When the intrinsic properties of an [=XR input source=] change for {{XRSession}} |session|, such as the {{XRInputSource/handedness}}, {{XRInputSource/targetRayMode}}, or presence of a {{XRInputSource/gripSpace}} or {{XRInputSource/gamepad}}, the user agent MUST run the following steps:
+When the intrinsic properties of any [=XR input source=]s change for {{XRSession}} |session|, such as the {{XRInputSource/handedness}}, {{XRInputSource/targetRayMode}}, or presence of a {{XRInputSource/gripSpace}} or {{XRInputSource/gamepad}}, the user agent MUST run the following steps:
```
I think I started that list thinking it was going to be an example then listed them all anyway and forgot to fix the start of the sentence. 😅

toji commented:
Update done with a sick, sleeping 2yo on my lap. Apologies from any oversights/typos as a result.

> I support the plural form of the event :)

Done!

> We haven't defined the expected behavior of the array when an item is added/removed.

I've updated the algorithms to use some [pre-defined list operations](https://infra.spec.whatwg.org/#list-extend), which should remove some of the ambiguity around how it should behave.

NellWaliczek commented on index.bs:
```diff
@@ -1229,7 +1232,7 @@ Gamepad API Integration {#gamepad-api-integration}
 
   - {{XRInputSource/gamepad}} MUST NOT be included in the array returned by {{navigator.getGamepads()}}.
   - {{XRInputSource/gamepad}}'s {{Gamepad/index}} attribute MUST be <code>0</code>.
```
>{{XRInputSource/gamepad}}'s {{Gamepad/index}} attribute MUST be <code>0</code>.

Gah this makes me so nervous.. people are going to use this and index into inputSources.... Can we file a separate issue (and label it to be an external dependency) to figure out how we want to handle this with the Gamepad spec folks? (yes, I'm looking at you @toji)

toji replied:
Re-cast #583 as a broader issue to address this topic.
